### ISSUE TITLE
When query contains double bracket `]]`, then it is read as a Optional Clause and querying fails

### ISSUE THREAD

**Issue by nfm** [metabase/metabase#11755]
**Describe the bug**
After upgrading from metabase 0.33.3 to 0.34.1, some saved questions that were created and working prior to the upgrade are now failing with an exception.

These questions contain a `{{date}}` variable. If I replace the variable with a string literal date (eg. `'2019-01-01'` the query runs successfully under 0.34.1.

The exception message is:

```
Output of parse-tokens* does not match schema: [0;33m [(named [nil nil (not (matches-some-precondition? a-clojure.lang.ChunkedCons))] "parsed tokens") (not (present? "remaining tokens"))] [0m 
```

**Logs**

Javascript console errors/warnings: none

Server logs: happy to share more if necessary but I've redacted the query data:

```
:class clojure.lang.ExceptionInfo,
:error
"Output of parse-tokens* does not match schema: \n\n\t [0;33m  [(named [nil nil (not (matches-some-precondition? a-clojure.lang.ChunkedCons))] \"parsed tokens\") (not (present? \"remaining tokens\"))] [0m \n\n",
:stacktrace
("--> driver.common.parameters.parse$fn__68708$parse_tokens_STAR___68713.invoke(parse.clj:84)"
"driver.common.parameters.parse$fn__68749$parse_tokens__68754$fn__68755.invoke(parse.clj:116)"
"driver.common.parameters.parse$fn__68749$parse_tokens__68754.invoke(parse.clj:114)"
"driver.common.parameters.parse$fn__68786$parse__68791$fn__68792.invoke(parse.clj:136)"
"driver.common.parameters.parse$fn__68786$parse__68791.invoke(parse.clj:127)"
"driver.sql$fn__71206.invokeStatic(sql.clj:42)"
"driver.sql$fn__71206.invoke(sql.clj:38)"
"query_processor.middleware.parameters.native$expand_inner.invokeStatic(native.clj:39)"
"query_processor.middleware.parameters.native$expand_inner.invoke(native.clj:30)"
"query_processor.middleware.parameters$expand_one.invokeStatic(parameters.clj:50)"
"query_processor.middleware.parameters$expand_one.invoke(parameters.clj:41)"
"query_processor.middleware.parameters$expand_all$replace_42698__42699.invoke(parameters.clj:59)"
"mbql.util.match$replace_in_collection$iter__24181__24185$fn__24186.invoke(match.clj:132)"
"mbql.util.match$replace_in_collection.invokeStatic(match.clj:131)"
"mbql.util.match$replace_in_collection.invoke(match.clj:126)"
"query_processor.middleware.parameters$expand_all$replace_42698__42699.invoke(parameters.clj:59)"
"query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:59)"
"query_processor.middleware.parameters$expand_all.invoke(parameters.clj:53)"
"query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:56)"
"query_processor.middleware.parameters$expand_all.invoke(parameters.clj:53)"
"query_processor.middleware.parameters$expand_parameters.invokeStatic(parameters.clj:77)"
"query_processor.middleware.parameters$expand_parameters.invoke(parameters.clj:73)"
"query_processor.middleware.parameters$fn__42714$substitute_parameters_STAR___42719$fn__42720.invoke(parameters.clj:82)"
"query_processor.middleware.parameters$fn__42714$substitute_parameters_STAR___42719.invoke(parameters.clj:79)"
"query_processor.middleware.driver_specific$process_query_in_context$fn__41596.invoke(driver_specific.clj:12)"
"query_processor.middleware.resolve_driver$resolve_driver$fn__43043.invoke(resolve_driver.clj:22)"
"query_processor.middleware.store$initialize_store$fn__46169$fn__46170.invoke(store.clj:11)"
"query_processor.store$do_with_store.invokeStatic(store.clj:46)"
"query_processor.store$do_with_store.invoke(store.clj:40)"
"query_processor.middleware.store$initialize_store$fn__46169.invoke(store.clj:10)"
"query_processor.middleware.async$async__GT_sync$fn__37384.invoke(async.clj:23)"
"query_processor.middleware.async_wait$runnable$fn__40637.invoke(async_wait.clj:89)"),
:query
{:native
{:template-tags
{:event_date {:id "2fc94564-2c94-1a1d-f37c-c32c54a82f52", :name "event_date", :display_name "Event date", :type "date", :display-name "Date", :default "2019-01-01", :required true}},
:query
<redacted>
:type "native",
:parameters [],
:async? true,
:middleware {:add-default-userland-constraints? true, :userland-query? true},
:info
{:executed-by 1,
:context :ad-hoc,
:card-id nil,
:nested? false,
:query-hash [-82, -25, -111, 91, -93, -79, -52, -46, -128, -41, 43, 49, -67, -74, -50, -70, 6, -94, 7, 72, 124, 83, 81, 78, -84, -128, -113, -27, 56, 17, -97, 40]},
:constraints {:max-results 10000, :max-results-bare-rows 2000}},
:ex-data
{:type :schema.core/error,
:value
<redacted>
:param-begin
"event_date"
:param-end
<redacted>
```

**To Reproduce**

I don't have a reproduction for this. I expect it relates to the specific query I'm using.

**Expected behavior**
Query that ran on 0.33.3 continues to run on 0.34.1.

**Screenshots**
None

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:72.0) Gecko/20100101 Firefox/72.0",
    "vendor": ""
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_232-heroku-b09",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_232-heroku",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.232-b09",
    "os.name": "Linux",
    "os.version": "4.4.0-1057-aws",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "heroku",
    "application-database": "postgres",
    "run-mode": "prod",
    "version": {
      "tag": "v0.34.1",
      "date": "2020-01-13",
      "branch": "release-0.34.x",
      "hash": "265695c"
    },
    "settings": {
      "report-timezone": "UTC"
    }
  }
}
```

**Severity**

Medium. Current workaround is to downgrade back to 0.33.3 but that won't be suitable long term.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by dacort** [comment]
Hi @nfm,

I'm having trouble reproducing this on my side. Are you able to reproduce this with a base installation or sample dataset and/or provide some more details on what exactly your native query is? There were [some changes](https://github.com/metabase/metabase/commit/71a58264b76987dd4428efe4c6446109e7091fee) there with newlines recently. I haven't tried that next, but will do so soon.

Here's what I tried so far using both the sample dataset and a Postgres application database:

1. Bring up 0.33.3
2. Create a new native query on the Sample Data
    ```sql
    SELECT * FROM orders WHERE CREATED_AT >= {{date}}
    ```
3. Configure the date variables a shown in screenshot below.
4. Restart Metabase with the new version (0.34.1), which will run any migrations
5. Try to access the saved question


![image](https://user-images.githubusercontent.com/1512/72761894-583edc00-3b92-11ea-9e90-fe2a9b2078fb.png)

**Comment by dacort** [comment]
Tried with a multi-line query as well, but no luck reproducing:

```sql
SELECT * FROM ORDERS
WHERE CREATED_AT >= {{date}}
ORDER BY CREATED_AT DESC
```

**Comment by nfm** [comment]
Thanks for taking a look @dacort :heart: I'll try to get back to you with a proper reproduction. I've got a pretty hectic week this week so it might not be until late next week. I'm guessing it's a whitespace thing, the queries I'm hitting this on are relatively long and nested.

**Comment by MelMarie80** [comment]
Hi - we had the same error message this morning after upgrading. I removed the comments in /* */ to resolve.

**Comment by dacort** [comment]
@MelMarie80 Do you have more details on what queries may have looked like before/after the upgrade?

I also tried with this SQL in v0.33.3 and upgraded to 0.34.1, but it still worked fine.

```sql
SELECT * FROM ORDERS
WHERE CREATED_AT >= {{date}}
/* this is a comment? */
ORDER BY CREATED_AT DESC
```

**Comment by nfm** [comment]
I just upgraded to v0.35.0-rc1 and ran into this again, and put in some time to come up with a minimal reproduction (based on the saved question we have that works on v0.33.3, but raises `Output of parse-tokens* does not match schema` on later versions).

This query fails for me on v0.35.0-rc1, but works as expected on v0.33.3:

```sql
SELECT SUBSTRING (
   'starting_plan: "test"'
   FROM
      '%starting_plan: #"[^[:space:]]*#"%' FOR '#'
)
where now() > {{date}}
```

![image](https://user-images.githubusercontent.com/127960/77617140-7f959900-6f87-11ea-8bac-5987a8c6b587.png)

![image](https://user-images.githubusercontent.com/127960/77617230-b8ce0900-6f87-11ea-85aa-4d9567896395.png)

It's some combination of the SQL regex and the `{{date}}` variable; if I change the where clause to `now() > now() - interval '1 day'` it works.

I appreciate this is probably pretty obscure :laughing:

**Comment by mjaadev** [comment]
Hello,

I am encountering the same issue with a similar query syntax: (the date variable is generating the error message mentioned below.

```sql
SELECT 
    strftime('%Y-%w', request.creation_date) "week",
	CASE 
        WHEN (UPPER(target_request.target) LIKE '%VALUE%' ) THEN 'Value'
        WHEN (UPPER(target_request.target) LIKE '*' ) THEN 'ALL'
        WHEN (UPPER(target_request.target) IN ('FAILURE', 'NOT_IDENTIFIED')) THEN 'OTHERS'
        Else target_request.target
    END "target value",
    COUNT(request.id) "count"
    -- 
FROM request
       LEFT OUTER JOIN target_request ON (request.id = target_request.request_id )
    -- 
WHERE 1=1
    AND (
    	UPPER(target_request.target) LIKE '%Value%' 
    	OR UPPER(target_request.target) LIKE '*'
    )
    -- AND request.creation_date >= '2020-02-01' AND request.creation_date < '2020-02-02'
    AND (request.creation_date >= {{start_date}} AND request.creation_date < {{end_date}})
GROUP BY 1, 2
ORDER BY 1 ASC;
```

> Error message: Output of parse-tokens* does not match schema: [0;33m [(named [nil nil (not (matches-some-precondition? a-clojure.lang.ChunkedCons))] "parsed tokens") (not (present? "remaining tokens"))] [0m

I supposed it was due to my filter with the "*" symbol but it is not. Same for the comment lines etc.... but it is not.
Do you have any idea why the date variable is generating the error?

**Comment by bitwisecook** [comment]
Here is my minimal repro:
```
select ot.text
from
  objtext ot
where
  ot.text ~ '[^\]]+'
;
```

if you change that to

```
select ot.text
from
  objtext ot
where
  ot.text ~ '[^\n]+'
;
```

then the error goes away.

**Comment by ericcj** [comment]
i ran across this as well when trying a regex on a bracket which is perfectly valid e.g.

regexp_replace(message, '\[[0-9]+\]', '#')

triggers "Invalid query: found [[ or {{ with no matching ]] or }}" and seems to imply it's a SQL error even though it's valid SQL if I run it directly:

<img width="401" alt="Screenshot 2024-07-29 at 2 28 33 PM" src="https://github.com/user-attachments/assets/9fe6511a-82ce-4261-a77e-a43c675cdd31">

So I had to add an optional space into my regex to work around it:

regexp_replace(message, '\[ ?[0-9]+\]', '#')

**Comment by paoliniluis** [comment]
@ericcj you say that this issue is not fixed?

**Comment by ericcj** [comment]
not as of 0.50.15 but i didn't expect it to be based on this ticket

**Comment by ericcj** [comment]
@snoe @paoliniluis this is not fixed in 0.50.20 still get "Invalid query: found [[ or {{ with no matching ]] or }}" if i include regexp_replace(message, '[[0-9]+]', '#') in the query.  i'm not sure if this is related to https://github.com/metabase/metabase/issues/46275 but that problem still exists as well thanks.

**Comment by ericcj** [comment]
oh the backport of #46547 (https://github.com/metabase/metabase/pull/46616) hasn't been merged yet i'll test again when you release that

**Comment by ericcj** [comment]
confirming this was fixed in 0.50.21 THANKS!
